ros/src/computing/planning/motion/packages/waypoint_follower/nodes/mpc_follower/mpc_utils.cpp

```diff
@@ -0,0 +1,478 @@
+
+// 1D interpolation
+bool MPCUtils::interp1dMPCTraj(const std::vector<double> &index, const MPCTrajectory &values,
+                               const std::vector<double> &ref_time, MPCTrajectory &ret)
+{
+  if (!(index.size() == values.size()))
+  {
+    printf("index and values must have same size, return false.\n");
+    return false;
+  }
+  if (index.size() == 1)
+  {
+    printf("index size is 1, too short. return false.\n");
+    return false;
+  }
+
+  for (unsigned int i = 1; i < index.size(); ++i)
+  {
+    if (!(index[i] > index[i - 1]))
+    {
+      printf("index must be monotonically increasing, return false. index[%d] = %f, but index[%d] = %f\n", i, index[i], i-1, index[i - 1]);
+      return false;
+    }
+  }
+
+  for (unsigned int i = 1; i < ref_time.size(); ++i)
+  {
+    if (!(ref_time[i] > ref_time[i - 1]))
+    {
+      printf("reference point must be monotonically increasing, return false. ref_time[%d] = %f, but ref_time[%d] = %f\n", i, ref_time[i], i-1, ref_time[i - 1]);
+      return false;
+    }
+  }
+
+  ret.clear();
+  unsigned int i = 1;
+  for (unsigned int j = 0; j < ref_time.size(); ++j)
+  {
+    double a, d_index;
+    if (ref_time[j] > index.back())
+    {
+      a = 1.0;
+      d_index = 1.0;
+      i = index.size() - 1;
+    }
+    else if (ref_time[j] < index.front())
+    {
+      a = 0.0;
+      d_index = 1.0;
+      i = 1;
+    }
+    else
+    {
+      while (ref_time[j] > index[i])
+      {
+        ++i;
+      }
+      a = ref_time[j] - index[i - 1];
+      d_index = index[i] - index[i - 1];
+    }
+    const double x = ((d_index - a) * values.x[i - 1] + a * values.x[i]) / d_index;
+    const double y = ((d_index - a) * values.y[i - 1] + a * values.y[i]) / d_index;
+    const double z = ((d_index - a) * values.z[i - 1] + a * values.z[i]) / d_index;
+    const double yaw = ((d_index - a) * values.yaw[i - 1] + a * values.yaw[i]) / d_index;
+    const double vx = ((d_index - a) * values.vx[i - 1] + a * values.vx[i]) / d_index;
+    const double k = ((d_index - a) * values.k[i - 1] + a * values.k[i]) / d_index;
+    const double t = ref_time[j];
+    ret.push_back(x, y, z, yaw, vx, k, t);
+  }
+  return true;
+}
+
+void MPCUtils::calcTrajectoryYawFromXY(MPCTrajectory &traj)
+{
+  if (traj.yaw.size() == 0)
+    return;
+
+  for (unsigned int i = 1; i < traj.yaw.size() - 1; ++i)
+  {
+    const double dx = traj.x[i + 1] - traj.x[i - 1];
+    const double dy = traj.y[i + 1] - traj.y[i - 1];
+    traj.yaw[i] = std::atan2(dy, dx);
+  }
+  if (traj.yaw.size() > 1)
+  {
+    traj.yaw[0] = traj.yaw[1];
+    traj.yaw.back() = traj.yaw[traj.yaw.size() - 2];
+  }
+}
+
+void MPCUtils::calcTrajectoryCurvature(MPCTrajectory &traj, int curvature_smoothing_num)
+{
+  unsigned int traj_k_size = traj.x.size();
+  traj.k.clear();
+
+  auto dist = [](const geometry_msgs::Point &a, const geometry_msgs::Point &b) {
```

Comment:
Please use implementation in [amathutils](https://github.com/autowarefoundation/autoware/blob/b734f194f9567953691be920378a8db3915592c8/ros/src/common/libs/amathutils_lib/include/amathutils_lib/amathutils.hpp#L54)

Reply:
Done. Thanks!